cmd/zfs/zfs_main.c

```diff
@@ -1122,11 +1123,22 @@ destroy_check_dependent(zfs_handle_t *zhp, void *data)
 	return (0);
 }
 
+static int
+destroy_batched(destroy_cbdata_t *cb)
+{
+	int error = zfs_destroy_snaps_nvl(g_zfs,
```

Comment:
This implementation seems, unfortunately, to violate the gentlemen's agreement of treating `zfs destroy -r` as an atomic operation:

https://github.com/zfsonlinux/zfs/blob/2705ebf0a74f15bb246fb899f19b0ee17aa1c4b5/lib/libzfs_core/libzfs_core.c#L304-L329

Consider the following (current) behaviour:

```
[root@linux ~]# POOLNAME='testpool'
[root@linux ~]# TMPDIR='/var/tmp'
[root@linux ~]# mountpoint -q $TMPDIR || mount -t tmpfs tmpfs $TMPDIR
[root@linux ~]# zpool destroy -f $POOLNAME
[root@linux ~]# fallocate -l 128m $TMPDIR/zpool.dat
[root@linux ~]# zpool create $POOLNAME $TMPDIR/zpool.dat
[root@linux ~]# zfs create $POOLNAME/fs
[root@linux ~]# for i in {1..20}; do
>    zfs snap $POOLNAME/fs@snap$i
> done
[root@linux ~]# zfs hold tag $POOLNAME/fs@snap15
[root@linux ~]# zfs destroy -vr $POOLNAME/fs
will destroy testpool/fs@snap14
will destroy testpool/fs@snap8
will destroy testpool/fs@snap2
will destroy testpool/fs@snap5
will destroy testpool/fs@snap19
will destroy testpool/fs@snap13
will destroy testpool/fs@snap1
will destroy testpool/fs@snap17
will destroy testpool/fs@snap10
will destroy testpool/fs@snap6
will destroy testpool/fs@snap4
will destroy testpool/fs@snap12
will destroy testpool/fs@snap18
will destroy testpool/fs@snap20
will destroy testpool/fs@snap15
will destroy testpool/fs@snap3
will destroy testpool/fs@snap9
will destroy testpool/fs@snap11
will destroy testpool/fs@snap7
will destroy testpool/fs@snap16
will destroy testpool/fs
cannot destroy snapshot testpool/fs@snap15: dataset is busy
[root@linux ~]# zfs list -t snapshot -r $POOLNAME/fs
NAME                 USED  AVAIL  REFER  MOUNTPOINT
testpool/fs@snap1      0B      -    24K  -
testpool/fs@snap2      0B      -    24K  -
testpool/fs@snap3      0B      -    24K  -
testpool/fs@snap4      0B      -    24K  -
testpool/fs@snap5      0B      -    24K  -
testpool/fs@snap6      0B      -    24K  -
testpool/fs@snap7      0B      -    24K  -
testpool/fs@snap8      0B      -    24K  -
testpool/fs@snap9      0B      -    24K  -
testpool/fs@snap10     0B      -    24K  -
testpool/fs@snap11     0B      -    24K  -
testpool/fs@snap12     0B      -    24K  -
testpool/fs@snap13     0B      -    24K  -
testpool/fs@snap14     0B      -    24K  -
testpool/fs@snap15     0B      -    24K  -
testpool/fs@snap16     0B      -    24K  -
testpool/fs@snap17     0B      -    24K  -
testpool/fs@snap18     0B      -    24K  -
testpool/fs@snap19     0B      -    24K  -
testpool/fs@snap20     0B      -    24K  -
[root@linux ~]# 
```

No snapshots were destroyed; now with this change:

```
[root@linux ~]# POOLNAME='testpool'
[root@linux ~]# TMPDIR='/var/tmp'
[root@linux ~]# mountpoint -q $TMPDIR || mount -t tmpfs tmpfs $TMPDIR
[root@linux ~]# zpool destroy -f $POOLNAME
[root@linux ~]# fallocate -l 128m $TMPDIR/zpool.dat
[root@linux ~]# zpool create $POOLNAME $TMPDIR/zpool.dat
[root@linux ~]# zfs create $POOLNAME/fs
[root@linux ~]# for i in {1..20}; do
>    zfs snap $POOLNAME/fs@snap$i
> done
[root@linux ~]# zfs hold tag $POOLNAME/fs@snap15
[root@linux ~]# cmd/zfs/zfs destroy -vr $POOLNAME/fs
will destroy testpool/fs@snap14
will destroy testpool/fs@snap19
will destroy testpool/fs@snap13
will destroy testpool/fs@snap7
will destroy testpool/fs@snap9
will destroy testpool/fs@snap3
will destroy testpool/fs@snap17
will destroy testpool/fs@snap10
will destroy testpool/fs@snap4
will destroy testpool/fs@snap12
will destroy testpool/fs@snap18
will destroy testpool/fs@snap20
will destroy testpool/fs@snap6
will destroy testpool/fs@snap1
will destroy testpool/fs@snap15
will destroy testpool/fs@snap11
will destroy testpool/fs@snap5
will destroy testpool/fs@snap2
will destroy testpool/fs@snap8
will destroy testpool/fs@snap16
cannot destroy snapshot testpool/fs@snap15: dataset is busy
will destroy testpool/fs
cannot destroy 'testpool/fs': dataset is busy
[root@linux ~]# zfs list -t snapshot -r $POOLNAME/fs
NAME                 USED  AVAIL  REFER  MOUNTPOINT
testpool/fs@snap1      0B      -    24K  -
testpool/fs@snap2      0B      -    24K  -
testpool/fs@snap5      0B      -    24K  -
testpool/fs@snap6      0B      -    24K  -
testpool/fs@snap8      0B      -    24K  -
testpool/fs@snap11     0B      -    24K  -
testpool/fs@snap15     0B      -    24K  -
testpool/fs@snap16     0B      -    24K  -
testpool/fs@snap18     0B      -    24K  -
testpool/fs@snap20     0B      -    24K  -
[root@linux ~]# 
```

Some of the snapshots were destroyed instead.

Reply:
it might make more sense to roll this into `destroy -d` behaviour or add a separate flag.

Reply:
> gentlemen's agreement of treating zfs destroy -r as an atomic operation:

I had the same thought initially, but I'm not sure that exactly true.  Please double check this for me, but it looks like `zfs destroy -r` _is_ currently only atomic in regard to all the snapshots for a single filesystem.  When recursively operating on multiple filesystems the entire operation is no longer atomic.

Reply:
@behlendorf yes, that exactly right, `lzc_destroy_snaps()` is atomic but `zfs destroy -r` is not because it does multiple calls to `lzc_destroy_snaps()` internally:

```
root@linux:~# POOLNAME='testpool'
root@linux:~# TMPDIR='/var/tmp'
root@linux:~# mountpoint -q $TMPDIR || mount -t tmpfs tmpfs $TMPDIR
root@linux:~# zpool destroy -f $POOLNAME
root@linux:~# rm -f $TMPDIR/zpool.dat
root@linux:~# fallocate -l 128m $TMPDIR/zpool.dat
root@linux:~# zpool create -f $POOLNAME $TMPDIR/zpool.dat
root@linux:~# zfs create $POOLNAME/fs
root@linux:~# zfs create $POOLNAME/fs/subfs1
root@linux:~# zfs create $POOLNAME/fs/subfs2
root@linux:~# for i in {1..3}; do
>    zfs snap -r $POOLNAME/fs@snap$i
> done
root@linux:~# zfs hold tag $POOLNAME/fs/subfs2@snap1
root@linux:~# stap -e '
> probe process("zfs").library("libzfs_core.so.1").function("lzc_destroy_snaps").call { printf(" -> %s %s\n", ppfunc(), $$parms$$); }
> probe process("zfs").library("libzfs_core.so.1").function("lzc_destroy_snaps").return { printf(" <- %s %s\n", ppfunc(), $$return$$); }
> ' -c "zfs destroy -vr $POOLNAME/fs"
will destroy testpool/fs/subfs1@snap2
will destroy testpool/fs/subfs1@snap1
will destroy testpool/fs/subfs1@snap3
will destroy testpool/fs/subfs1
will destroy testpool/fs/subfs2@snap3
will destroy testpool/fs/subfs2@snap1
will destroy testpool/fs/subfs2@snap2
will destroy testpool/fs/subfs2
cannot destroy snapshot testpool/fs/subfs2@snap1: dataset is busy
WARNING: Child process exited with status 1
 -> lzc_destroy_snaps snaps={.nvl_version=0, .nvl_nvflag=1, .nvl_priv=34380064, .nvl_flag=0, .nvl_pad=0} defer=0 errlist=ERROR
 <- lzc_destroy_snaps return=0
 -> lzc_destroy_snaps snaps={.nvl_version=0, .nvl_nvflag=1, .nvl_priv=34380064, .nvl_flag=0, .nvl_pad=0} defer=0 errlist=ERROR
 <- lzc_destroy_snaps return=16
WARNING: /usr/bin/staprun exited with status: 1
Pass 5: run failed.  [man error::pass5]
root@linux:~# zfs list -t snap
NAME                       USED  AVAIL  REFER  MOUNTPOINT
testpool/fs@snap1            0B      -    24K  -
testpool/fs@snap2            0B      -    24K  -
testpool/fs@snap3            0B      -    24K  -
testpool/fs/subfs2@snap1     0B      -    24K  -
testpool/fs/subfs2@snap2     0B      -    24K  -
testpool/fs/subfs2@snap3     0B      -    24K  -
root@linux:~# 
```

The first call to `lzc_destroy_snaps()` succeeds, the seconds does not:

```
 -> lzc_destroy_snaps snaps={.nvl_version=0, .nvl_nvflag=1, .nvl_priv=34380064, .nvl_flag=0, .nvl_pad=0} defer=0 errlist=ERROR
 <- lzc_destroy_snaps return=0
 -> lzc_destroy_snaps snaps={.nvl_version=0, .nvl_nvflag=1, .nvl_priv=34380064, .nvl_flag=0, .nvl_pad=0} defer=0 errlist=ERROR
 <- lzc_destroy_snaps return=16
```

There may not be a real issue here, maybe `zfs destroy -r` has only been "coincidentally atomic" on single datasets due to the specific implementation and it is consequently safe to change the behaviour, especially if said behaviour is not documented anywhere.

Reply:
I'm trying to think of situations where the user might actually expect this (apparently undocumented) behaviour, though there is always an argument to be made that you should not expect consistency if the interface does not guarantee it. Use libzfs_core instead!

```
root@linux:~# POOLNAME='testpool'
root@linux:~# TMPDIR='/var/tmp'
root@linux:~# mountpoint -q $TMPDIR || mount -t tmpfs tmpfs $TMPDIR
root@linux:~# zpool destroy -f $POOLNAME
root@linux:~# rm -f $TMPDIR/zpool.dat
root@linux:~# fallocate -l 128m $TMPDIR/zpool.dat
root@linux:~# zpool create -f $POOLNAME $TMPDIR/zpool.dat
root@linux:~# zfs create $POOLNAME/fs
root@linux:~# zfs create $POOLNAME/fs/subfs1
root@linux:~# zfs create $POOLNAME/fs/subfs2
root@linux:~# for i in {1..3}; do
>    zfs snap -r $POOLNAME/fs@snap$i
> done
root@linux:~# zfs hold tag $POOLNAME/fs/subfs2@snap1
root@linux:~# 
root@linux:~# zfs list -t snap | wc -l
10
root@linux:~# python
Python 2.7.9 (default, Jun 29 2016, 13:08:31) 
[GCC 4.9.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import subprocess
>>> import libzfs_core as lzc
>>> 
>>> snaps = subprocess.check_output("zfs list -t snap -o name -H", shell=True).split()
>>> try:
...  lzc.lzc_destroy_snaps(snaps, False)
... except lzc.exceptions.SnapshotDestructionFailure as e:
...  print e.errors
... 
[SnapshotIsHeld(16, 'Snapshot is held')]
>>> 
root@linux:~# zfs list -t snap | wc -l
10
root@linux:~# 
```